application/app.py

```diff
@@ -93,28 +95,62 @@ def get_messages():
 
 @app.route("/api/message", methods=['POST'])
 def incoming_message():
+    """ Respond to an incoming SMS message.
+
+        First, checks to see if the message can be recognized as one of the
+        commands in the SMS API and responds appropriately if so. If not, this
+        will try to determine whether the message is in response to another
+        message or in a current event, in order to determine how to thread the
+        message.  """
     from_number = request.values.get('From', None)
     body = request.values.get('Body', None)
     timestamp = int(time.time())
     user = User.from_number(from_number)
     event = 0
+    # The ID of the last message sent to the user is stored on the user object.
+    # If that exists and refers to an active event, we assume this message
+    # related to the samne event.
     if user.last_msg:
         last_msg = Message.query.get(user.last_msg)
-        event = last_msg.event
-
-    message = Message(
-        text=body, 
-        author=user.id, 
-        timestamp=timestamp, 
-        parent=user.last_msg, 
-        event=event
-        )
-    db.session.add(message)
-    db.session.commit()
+        if last_msg.event:
+            """ If the event isn't active any more, consideer this a new message. """
+            e = Event.query.get(last_msg.event)
+            if e and e.active:
+                event = last_msg_event.id
+            else:
+                user.last_msg = 0
+                event = 0
+
+
+    # Parse the message for special commands (see MsgParse class)
+    msg_handler = parse_message(body)
+    if msg_handler['handled']:
+        user_actions = UserActions(
+                    user=user,
+                    event=event,
+                    timestamp=timestamp,
+                    )
+        response = msg_handler['response'] 
+        action = getattr(user_actions, msg_handler['action'], None)
+        if callable(action):
+            action()
```

Comment:
Part of me is always a little worried when I see function calls based on user input, but after digging through this, I think this is ok as we are still using the data from MsgParse to give the name of the function to call.

Reply:
Yeah, I _think_ this should be OK, since this can only call functions that are defined in [UserActions](https://github.com/PortlandAssembly/The-project/blob/7b584aef430d619f6fe2894bcd930049e1aabfcb/application/utils/useractions.py), and those functions should be responsible for checking authorization for anything that requires it.